File: BaseLibrary/NetworkStream.cs
using System;
using System.Collections.Concurrent;
using System.ComponentModel.Design;
using System.IO;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using BaseLibrary.command;
using BaseLibrary.command.handshake;
using BaseLibrary.protocol;

namespace BaseLibrary {
    /// <summary>
    /// Stream for communication by sockets with server.
    /// </summary>
	public class NetworkStream {

		private static int ID = 0;

		private int id = ID++;

		protected readonly System.Net.Sockets.NetworkStream ns;
		protected readonly StreamReader sr;
		protected readonly StreamWriter sw;
		protected AProtocol _protocol = null;
		public AProtocol PROTOCOL {
			get {
				return _protocol;
			}
			set {
				if (_protocol == null || _protocol == value) {
					_protocol = value;
				} else {
					throw new Exception("Protocol can be set only once.");
				}
			}
		}

        /// <summary>
        /// Create instance from socket.
        /// </summary>
        /// <param name="s"></param>
		public NetworkStream(Socket s) {
			ns = new System.Net.Sockets.NetworkStream(s, true);
			sr = new StreamReader(ns, System.Text.Encoding.UTF8);
			sw = new StreamWriter(ns, System.Text.Encoding.UTF8) {
				NewLine = "\n",
				AutoFlush = true
			};
		}

        /// <summary>
        /// Read asynchronously one line from stream.
        /// </summary>
        /// <returns></returns>
		public async Task<string> ReadLineAsync() {
			return await sr.ReadLineAsync();
		}

        /// <summary>
        /// Write asynchronously one line to stream.
        /// </summary>
        /// <param name="s"></param>
        /// <returns></returns>
		public async Task WriteLineAsync(string s) {
		    try {
		        await sw.WriteLineAsync(s);
		    } catch (IOException) {
		        // socket was closed
                this.Close();
		    }
		}

        /// <summary>
        /// Synchronously receive command from stream. That is mean read line and deserialize command by <code>PROTOCOL</code>.
        /// </summary>
        /// <exception cref="ArgumentException">When protocol can not deserialize line.</exception>
        /// <returns></returns>
		public virtual ACommand ReceiveCommand() {
		    return ReceiveCommandAsync().Result;
		}

        /// <summary>
        /// Asynchronously receive command from stream. That is mean read line and deserialize command by <code>PROTOCOL</code>.
        /// </summary>
        /// <exception cref="ArgumentException">When protocol can not deserialize line.</exception>
        /// <returns></returns>
	    public virtual async Task<ACommand> ReceiveCommandAsync() {
	        if (_protocol == null) {
	            throw new Exception("Cannot read or write before set protocol.");
	        }

	        string s = await sr.ReadLineAsync();
	        ACommand command = PROTOCOL.GetCommand(s);
	        if (command == null) {
	            throw new ArgumentException("Protocol (" + _protocol.GetType().Name + ") can not deserialize string - " +
	                                        s);
	        }

	        if (command is ErrorCommand) {
	            Console.Error.WriteLine(((ErrorCommand) command).MESSAGE);
	        }
	        return command;
	    }

        /// <summary>
        /// Synchronously send command.
        /// </summary>
        /// <param name="command"></param>
	    public virtual void SendCommand(ACommand command) {
	        SendCommandAsync(command).Wait();
	    }

        /// <summary>
        /// Asynchronously send command.
        /// </summary>
        /// <param name="command"></param>
        /// <returns></returns>
        public virtual async Task SendCommandAsync(ACommand command) {
			if (_protocol == null) {
				throw new Exception("Cannot read or write before set protocol.");
			}

            string serializedCommand = PROTOCOL.GetSendableCommand(command);

            await WriteLineAsync(serializedCommand);
		}

	    private bool close = false;

        /// <summary>
        /// Close stream.
        /// </summary>
		public void Close() {
		    if (!close) {
		        sw.Dispose();
		        sr.Dispose();
		        ns.Close();
		        close = true;
		    }
		}
	}
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/communication/NetworkStream.java
package BaseLibrary.communication;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

import BaseLibrary.communication.command.ACommand;
import BaseLibrary.communication.command.handshake.ErrorCommand;
import BaseLibrary.communication.protocol.AProtocol;

/**
 * Stream for communication by sockets with server.
 */
public class NetworkStream {

	protected final Socket s;
	protected final Reader sr;
	protected final Writer sw;
	protected AProtocol protocol = null;

	public AProtocol getProtocol() {
		return protocol;
	}

	public void setProtocol(AProtocol protocol) {
		if (this.protocol == null || this.protocol.equals(protocol)) {
			this.protocol = protocol;
		} else {
			throw new RuntimeException("Protocol can be set only once.");
		}
	}

	/**
	 * Create instance from socket.
	 * 
	 * @throws IOException
	 */
	public NetworkStream(Socket s) throws IOException {
		this.s = s;
		sr = new BufferedReader(new InputStreamReader(s.getInputStream(), StandardCharsets.UTF_8));
		sw = new BufferedWriter(new OutputStreamWriter(s.getOutputStream(), StandardCharsets.UTF_8)) {
			@Override
			public void newLine() throws IOException {
				this.write('\n');
			}
		};

	}

	/**
	 * Read asynchronously one line from stream.
	 */
	public CompletableFuture<String> readLineAsync() {
		return CompletableFuture.supplyAsync(() -> {
			try {
				StringBuilder s = new StringBuilder();
				int character = 0;
				while ((character = sr.read()) != -1 && character != '\n') {
					if (character == Integer.parseInt("FEFF", 16)) {
						continue;
					}
					s.append((char) character);
				}
				return s.toString();
			} catch (IOException e) {
				throw new CompletionException(e);
			}
		});
	}

	/**
	 * Write asynchronously one line to stream.
	 */
	public CompletableFuture<Void> writeLineAsync(String s) {
		return CompletableFuture.runAsync(() -> {
			try {
				sw.write(s);
				sw.write('\n');
				sw.flush();
			} catch (IOException e) {
				throw new CompletionException(e.getMessage(), e);
			}
		});
	}

	/**
	 * Synchronously receive command from stream. That is mean read line and
	 * deserialize command by <code>PROTOCOL</code>.
	 * 
	 * @throws IllegalArgumentException
	 *             - When protocol can not deserialize line.
	 */
	public ACommand receiveCommand() throws IllegalArgumentException {
		return receiveCommandAsync().join();
	}

	/**
	 * Asynchronously receive command from stream. That is mean read line and
	 * deserialize command by <code>PROTOCOL</code>.
	 * 
	 * @throws RuntimeException
	 *             - When protocol can not deserialize line. - When protocol is not
	 *             set.
	 */
	public CompletableFuture<ACommand> receiveCommandAsync() {
		if (protocol == null) {
			throw new RuntimeException("Cannot read or write before set protocol.");
		}

		return readLineAsync().thenApplyAsync((s) -> {
				System.out.println(s);
				ACommand command = getProtocol().getCommand(s);
				
				if (command == null) {
					throw new RuntimeException(String.format("Protocol (%s) can not deserialize String '%s'",
							protocol.getClass().getName(), s));
				}

				if (command instanceof ErrorCommand) {
					System.err.println(((ErrorCommand) command).MESSAGE);
				}
				return command;	
			});
	}

	/**
	 * Synchronously send command.
	 */
	public void sendCommand(ACommand command) {
		sendCommandAsync(command).join();
	}

	/**
	 * Asynchronously send command.
	 */
	public CompletableFuture<Void> sendCommandAsync(ACommand command) {
		if (protocol == null) {
			throw new RuntimeException("Cannot read or write before set protocol.");
		}

		String serializedCommand = command.serialize(getProtocol());
		return writeLineAsync(serializedCommand);
	}

	private boolean close = false;

	/**
	 * Close stream.
	 */
	public void close() {
		if (!close) {
			try {
				sw.close();
			} catch (IOException e) {
			}
			try {
				sr.close();
			} catch (IOException e) {
			}
			try {
				s.close();
			} catch (IOException e) {
			}
			close = true;
		}
	}
}